nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -529,4 +542,525 @@ impl DataStore {
+                "instance_id" => %instance_id,
+                "saga_id" => %saga_lock_id,
+                "current_gen" => ?current_gen,
+            );
+
+            (instance, did_lock) = diesel::update(dsl::instance)
+                .filter(dsl::time_deleted.is_null())
+                .filter(dsl::id.eq(instance_id))
+                // If the generation is the same as the captured generation when we
+                // read the instance record to check if it was not locked, we can
+                // lock this instance. This is because changing the `updater_id`
+                // field always increments the generation number. Therefore, we
+                // want the update query to succeed if and only if the
+                // generation number remains the same as the generation when we
+                // last fetched the instance. This query is used equivalently to
+                // an atomic compare-and-swap instruction in the implementation
+                // of a non-distributed, single-process mutex.
+                .filter(dsl::updater_gen.eq(current_gen))
+                .set((
+                    dsl::updater_gen.eq(dsl::updater_gen + 1),
+                    dsl::updater_id.eq(Some(saga_lock_id)),
+                ))
+                .check_if_exists::<Instance>(instance_id)
+                .execute_and_check(
+                    &*self.pool_connection_authorized(opctx).await?,
+                )
+                .await
+                .map(|r| {
+                    // If we successfully updated the instance record, we have
+                    // acquired the lock; otherwise, we haven't --- either because
+                    // our generation is stale, or because the instance is already locked.
+                    let locked = match r.status {
+                        UpdateStatus::Updated => true,
+                        UpdateStatus::NotUpdatedButExists => false,
+                    };
+                    (r.found, locked)
+                })
+                .map_err(|e| {
+                    public_error_from_diesel(
+                        e,
+                        ErrorHandler::NotFoundByLookup(
+                            ResourceType::Instance,
+                            LookupType::ById(instance_id),
+                        ),
+                    )
+                })?;
+        }
+    }
+
+    /// Release the instance-updater lock acquired by
+    /// [`DataStore::instance_updater_lock`].
+    ///
+    /// This method will unlock the instance if (and only if) the lock is
+    /// currently held by the provided `saga_lock_id`. If the lock is held by a
+    /// different saga UUID, the instance will remain locked. If the instance
+    /// has already been unlocked, this method will return `false`.
+    ///
+    /// # Arguments
+    ///
+    /// - `authz_instance`: the instance to attempt to unlock
+    /// - `current_gen`: the generation number for the lock being released.
```

Comment:
I would possibly rename this to something like `after_lock_acquired_gen` and note explicitly that it needs to be the lock generation number that was produced by acquiring the lock. (For many sagas I expect it will be very easy to reach for the generation number that was in the instance record when the saga started.)

Maybe we should leverage the type system here: have `instance_updater_lock` return an opaque struct that contains the lock ID and the post-acquisition generation and make `instance_updater_unlock` take that type. Harder to mishandle the values that way. WDYT?

Reply:
> Maybe we should leverage the type system here: have `instance_updater_lock` return an opaque struct that contains the lock ID and the post-acquisition generation and make `instance_updater_unlock` take that type. Harder to mishandle the values that way. WDYT?

I actually did this, originally, on the other branch! I removed it because I wasn't sure if a saga node's reverse action would be able to access the output of its forward action, so I wasn't sure if there was a way for the "release lock on unwind" reverse action to get the opaque token or not. If that's possible, I'll definitely try this approach.

Reply:
> I removed it because I wasn't sure if a saga node's reverse action would be able to access the output of its forward action

I believe you can; see e.g. the `CREATE_VMM_RECORD` step in the instance start saga, where `sis_destroy_vmm_record` refers to the `vmm_record` output from `sis_create_vmm_record`.

Reply:
Ah, wonderful. In that case, I'll see about making this change.